File: Kikaria-Windows/src/Kikaria.Core/MathFallback.cs
//
//  MathFallback.cs
//  Kikaria-Windows
//
//  readableMathFallback 的完整移植,来自 Kikaria-Apple 的 KikariaMathFormulaView.swift:
//  - \frac{a}{b} → "(a) / (b)"(分量含空格/换行/斜杠时加括号,递归处理);
//  - \sqrt{x} → "√(x)";
//  - 剥 \operatorname / \mathrm / \mathbf / \text(\bar 追加组合上横线 U+0304);
//  - \begin{cases} 等环境 → 空、\\ → 换行、& → 两空格、\quad → 空格;
//  - \left/\right/\Bigg/\bigg/\Big/\big → 空;
//  - 希腊字母与运算符全表逐条照抄源文件。
//  替换顺序必须保持与源一致(长命令在前,如 \iiint → \iint → \int)。
//

using System.Text;

namespace Kikaria.Core;

public static class MathFallback
{
    private static readonly (string Source, string Replacement)[] Replacements =
    {
        ("\\begin{cases}", ""),
        ("\\end{cases}", ""),
        ("\\begin{aligned}", ""),
        ("\\end{aligned}", ""),
        ("\\begin{matrix}", ""),
        ("\\end{matrix}", ""),
        ("\\\\", "\n"),
        ("&", "  "),
        ("\\qquad", "  "),
        ("\\quad", " "),
        ("\\,", " "),
        ("\\;", " "),
        ("\\:", " "),
        ("\\!", ""),
        ("\\left", ""),
        ("\\right", ""),
        ("\\Bigg", ""),
        ("\\bigg", ""),
        ("\\Big", ""),
        ("\\big", ""),
        ("\\iiint", "∫∫∫"),
        ("\\iint", "∫∫"),
        ("\\int", "∫"),
        ("\\partial", "∂"),
        ("\\nabla", "∇"),
        ("\\Delta", "Δ"),
        ("\\delta", "δ"),
        ("\\theta", "θ"),
        ("\\rho", "ρ"),
        ("\\xi", "ξ"),
        ("\\alpha", "α"),
        ("\\beta", "β"),
        ("\\gamma", "γ"),
        ("\\lambda", "λ"),
        ("\\mu", "μ"),
        ("\\pi", "π"),
        ("\\neq", "≠"),
        ("\\ne", "≠"),
        ("\\leq", "≤"),
        ("\\le", "≤"),
        ("\\geq", "≥"),
        ("\\ge", "≥"),
        ("\\to", "→"),
        ("\\infty", "∞"),
        ("\\cdots", "⋯"),
        ("\\ldots", "…"),
        ("\\times", "×"),
        ("\\pm", "±"),
        ("\\mp", "∓"),
        ("\\in", "∈")
    };

    /// <summary>块级公式的可读文本(等价于 Apple 版 readableMathFallback)。</summary>
    public static string Readable(string source)
    {
        var readable = StripBlockMathDelimiters(source);
        readable = ReplaceFractions(readable);
        readable = ReplaceBracedCommand(readable, "\\sqrt", value => "√(" + Readable(value) + ")");
        readable = ReplaceBracedCommand(readable, "\\operatorname", value => value);
        readable = ReplaceBracedCommand(readable, "\\mathrm", value => value);
        readable = ReplaceBracedCommand(readable, "\\mathbf", value => value);
        readable = ReplaceBracedCommand(readable, "\\text", value => value);
        readable = ReplaceBracedCommand(readable, "\\bar", value => Readable(value) + "\u0304");

        foreach (var (source1, replacement) in Replacements)
        {
            readable = readable.Replace(source1, replacement);
        }

        readable = readable.Replace("\\", "");
        return NormalizeFallbackWhitespace(readable);
    }

    /// <summary>行内公式的可读文本:去掉 $ 定界符后同样走可读化转换。</summary>
    public static string InlineReadable(string body) => Readable(body);

    private static string StripBlockMathDelimiters(string source)
    {
        var trimmed = source.Trim();
        if (trimmed.StartsWith("$$", StringComparison.Ordinal) &&
            trimmed.EndsWith("$$", StringComparison.Ordinal) &&
            trimmed.Length >= 4)
        {
            return trimmed.Substring(2, trimmed.Length - 4).Trim();
        }

        return trimmed;
    }

    private static string ReplaceFractions(string text)
    {
        return ReplaceBracedPairCommand(text, new[] { "\\dfrac", "\\tfrac", "\\frac" }, (numerator, denominator) =>
        {
            var readableNumerator = Readable(numerator);
            var readableDenominator = Readable(denominator);
            return WrapComponent(readableNumerator) + " / " + WrapComponent(readableDenominator);
        });
    }

    private static string ReplaceBracedPairCommand(
        string text,
        string[] commands,
        Func<string, string, string> transform)
    {
        var result = new StringBuilder();
        var index = 0;

        while (index < text.Length)
        {
            var matched = false;
            foreach (var command in commands)
            {
                if (!HasPrefixAt(text, command, index))
                {
                    continue;
                }

                var cursor = index + command.Length;
                SkipWhitespace(text, ref cursor);

                if (BracedGroup(text, ref cursor, out var firstValue))
                {
                    SkipWhitespace(text, ref cursor);
                    if (BracedGroup(text, ref cursor, out var secondValue))
                    {
                        result.Append(transform(firstValue, secondValue));
                        index = cursor;
                        matched = true;
                        break;
                    }
                }
            }

            if (matched)
            {
                continue;
            }

            result.Append(text[index]);
            index++;
        }

        return result.ToString();
    }

    private static string ReplaceBracedCommand(string text, string command, Func<string, string> transform)
    {
        var result = new StringBuilder();
        var index = 0;

        while (index < text.Length)
        {
            if (HasPrefixAt(text, command, index))
            {
                var cursor = index + command.Length;
                SkipWhitespace(text, ref cursor);

                if (BracedGroup(text, ref cursor, out var value))
                {
                    result.Append(transform(value));
                    index = cursor;
                    continue;
                }
            }

            result.Append(text[index]);
            index++;
        }

        return result.ToString();
    }

    private static bool BracedGroup(string text, ref int cursor, out string value)
    {
        value = string.Empty;
        if (cursor >= text.Length || text[cursor] != '{')
        {
            return false;
        }

        cursor++;
        var start = cursor;
        var depth = 1;

        while (cursor < text.Length)
        {
            if (text[cursor] == '{')
            {
                depth++;
            }
            else if (text[cursor] == '}')
            {
                depth--;
                if (depth == 0)
                {
                    value = text.Substring(start, cursor - start);
                    cursor++;
                    return true;
                }
            }

            cursor++;
        }

        return false;
    }

    private static void SkipWhitespace(string text, ref int cursor)
    {
        while (cursor < text.Length && char.IsWhiteSpace(text[cursor]))
        {
            cursor++;
        }
    }

    private static string WrapComponent(string value)
    {
        if (value.Contains(' ') || value.Contains('\n') || value.Contains('/'))
        {
            return "(" + value + ")";
        }

        return value;
    }

    private static string NormalizeFallbackWhitespace(string source)
    {
        var normalizedLines = new List<string>();
        foreach (var rawLine in source.Split('\n'))
        {
            var line = rawLine.Trim();
            while (line.Contains("  ", StringComparison.Ordinal))
            {
                line = line.Replace("  ", " ");
            }

            if (line.Length == 0)
            {
                if (normalizedLines.Count > 0 && normalizedLines[^1].Length != 0)
                {
                    normalizedLines.Add(line);
                }
            }
            else
            {
                normalizedLines.Add(line);
            }
        }

        return string.Join("\n", normalizedLines);
    }

    private static bool HasPrefixAt(string text, string prefix, int index)
    {
        if (index < 0 || index + prefix.Length > text.Length)
        {
            return false;
        }

        for (var offset = 0; offset < prefix.Length; offset++)
        {
            if (text[index + offset] != prefix[offset])
            {
                return false;
            }
        }

        return true;
    }
}

File: Kikaria-Android/app/src/main/java/com/vita0818/kikaria/math/MathFallback.kt
package com.vita0818.kikaria.math

/**
 * 公式渲染失败时的可读文本转换,逐条对齐 KikariaMathFormulaView.swift 的 readableMathFallback。
 */
object MathFallback {

    fun readableMathFallback(source: String): String {
        var readable = strippedBlockMathDelimiters(source)
        readable = replacingFractions(readable)
        readable = replacingBracedCommand(readable, "\\sqrt") { "√(${readableMathFallback(it)})" }
        readable = replacingBracedCommand(readable, "\\operatorname") { it }
        readable = replacingBracedCommand(readable, "\\mathrm") { it }
        readable = replacingBracedCommand(readable, "\\mathbf") { it }
        readable = replacingBracedCommand(readable, "\\text") { it }
        readable = replacingBracedCommand(readable, "\\bar") { "${readableMathFallback(it)}\u0304" }

        val replacements = listOf(
            "\\begin{cases}" to "", "\\end{cases}" to "",
            "\\begin{aligned}" to "", "\\end{aligned}" to "",
            "\\begin{matrix}" to "", "\\end{matrix}" to "",
            "\\\\" to "\n", "&" to "  ",
            "\\qquad" to "  ", "\\quad" to " ",
            "\\," to " ", "\\;" to " ", "\\:" to " ", "\\!" to "",
            "\\left" to "", "\\right" to "",
            "\\Bigg" to "", "\\bigg" to "", "\\Big" to "", "\\big" to "",
            "\\iiint" to "∫∫∫", "\\iint" to "∫∫", "\\int" to "∫",
            "\\partial" to "∂", "\\nabla" to "∇",
            "\\Delta" to "Δ", "\\delta" to "δ", "\\theta" to "θ", "\\rho" to "ρ",
            "\\xi" to "ξ", "\\alpha" to "α", "\\beta" to "β", "\\gamma" to "γ",
            "\\lambda" to "λ", "\\mu" to "μ", "\\pi" to "π",
            "\\neq" to "≠", "\\ne" to "≠",
            "\\leq" to "≤", "\\le" to "≤", "\\geq" to "≥", "\\ge" to "≥",
            "\\to" to "→", "\\infty" to "∞",
            "\\cdots" to "⋯", "\\ldots" to "…", "\\times" to "×",
            "\\pm" to "±", "\\mp" to "∓", "\\in" to "∈",
        )
        for ((from, to) in replacements) readable = readable.replace(from, to)

        readable = readable.replace("\\", "")
        return normalizedFallbackWhitespace(readable)
    }

    fun strippedBlockMathDelimiters(source: String): String {
        val trimmed = source.trim()
        if (trimmed.startsWith("$$") && trimmed.endsWith("$$") && trimmed.length >= 4) {
            return trimmed.drop(2).dropLast(2).trim()
        }
        return trimmed
    }

    private fun replacingFractions(text: String): String =
        replacingBracedPairCommand(text, listOf("\\dfrac", "\\tfrac", "\\frac")) { num, den ->
            val n = readableMathFallback(num)
            val d = readableMathFallback(den)
            "${wrappedMathFallbackComponent(n)} / ${wrappedMathFallbackComponent(d)}"
        }

    private fun replacingBracedPairCommand(
        text: String,
        commands: List<String>,
        transform: (String, String) -> String,
    ): String {
        val sb = StringBuilder()
        var i = 0
        while (i < text.length) {
            val command = commands.firstOrNull { text.startsWith(it, i) }
            if (command != null) {
                var cursor = skipWhitespace(text, i + command.length)
                val first = bracedGroup(text, cursor)
                if (first != null) {
                    cursor = skipWhitespace(text, first.second)
                    val second = bracedGroup(text, cursor)
                    if (second != null) {
                        sb.append(transform(first.first, second.first))
                        i = second.second
                        continue
                    }
                }
            }
            sb.append(text[i])
            i += 1
        }
        return sb.toString()
    }

    private fun replacingBracedCommand(text: String, command: String, transform: (String) -> String): String {
        val sb = StringBuilder()
        var i = 0
        while (i < text.length) {
            if (text.startsWith(command, i)) {
                val cursor = skipWhitespace(text, i + command.length)
                val group = bracedGroup(text, cursor)
                if (group != null) {
                    sb.append(transform(group.first))
                    i = group.second
                    continue
                }
            }
            sb.append(text[i])
            i += 1
        }
        return sb.toString()
    }

    /** 返回 (组内容, 结束后光标);不是 { 开头时返回 null。 */
    private fun bracedGroup(text: String, start: Int): Pair<String, Int>? {
        if (start >= text.length || text[start] != '{') return null
        var cursor = start + 1
        val begin = cursor
        var depth = 1
        while (cursor < text.length) {
            when (text[cursor]) {
                '{' -> depth += 1
                '}' -> {
                    depth -= 1
                    if (depth == 0) return text.substring(begin, cursor) to (cursor + 1)
                }
            }
            cursor += 1
        }
        return null
    }

    private fun skipWhitespace(text: String, from: Int): Int {
        var i = from
        while (i < text.length && text[i].isWhitespace()) i += 1
        return i
    }

    private fun wrappedMathFallbackComponent(value: String): String =
        if (value.contains(' ') || value.contains('\n') || value.contains('/')) "($value)" else value

    private fun normalizedFallbackWhitespace(source: String): String {
        val lines = mutableListOf<String>()
        source.split("\n").forEach { raw ->
            var line = raw.trim()
            while (line.contains("  ")) line = line.replace("  ", " ")
            if (line.isEmpty()) {
                if (lines.lastOrNull()?.isEmpty() == false) lines.add(line)
            } else {
                lines.add(line)
            }
        }
        return lines.joinToString("\n")
    }
}
